sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_receiver.py

```diff
@@ -363,68 +371,77 @@ def _open(self):
 
+    def _enhanced_message_received(self, message):
+        if self._receive_context.is_set():
+            self._handler._received_messages.put(message)
+        else:
+            message.release()
+
     def _receive(self, max_message_count=None, timeout=None):
         # type: (Optional[int], Optional[float]) -> List[ServiceBusReceivedMessage]
         # pylint: disable=protected-access
-        self._open()
-
-        amqp_receive_client = self._handler
-        received_messages_queue = amqp_receive_client._received_messages
-        max_message_count = max_message_count or self._prefetch_count
-        timeout_ms = (
-            1000 * (timeout or self._max_wait_time)
-            if (timeout or self._max_wait_time)
-            else 0
-        )
-        abs_timeout_ms = (
-            amqp_receive_client._counter.get_current_ms() + timeout_ms
-            if timeout_ms
-            else 0
-        )
-
-        batch = []  # type: List[Message]
-        while not received_messages_queue.empty() and len(batch) < max_message_count:
-            batch.append(received_messages_queue.get())
-            received_messages_queue.task_done()
-        if len(batch) >= max_message_count:
-            return [self._build_message(message) for message in batch]
-
-        # Dynamically issue link credit if max_message_count > 1 when the prefetch_count is the default value 1
-        if max_message_count and self._prefetch_count == 1 and max_message_count > 1:
-            link_credit_needed = max_message_count - len(batch)
-            amqp_receive_client.message_handler.reset_link_credit(link_credit_needed)
-
-        first_message_received = expired = False
-        receiving = True
-        while receiving and not expired and len(batch) < max_message_count:
-            while receiving and received_messages_queue.qsize() < max_message_count:
-                if (
-                    abs_timeout_ms
-                    and amqp_receive_client._counter.get_current_ms() > abs_timeout_ms
-                ):
-                    expired = True
-                    break
-                before = received_messages_queue.qsize()
-                receiving = amqp_receive_client.do_work()
-                received = received_messages_queue.qsize() - before
-                if (
-                    not first_message_received
-                    and received_messages_queue.qsize() > 0
-                    and received > 0
-                ):
-                    # first message(s) received, continue receiving for some time
-                    first_message_received = True
-                    abs_timeout_ms = (
-                        amqp_receive_client._counter.get_current_ms()
-                        + self._further_pull_receive_timeout_ms
-                    )
-            while (
-                not received_messages_queue.empty() and len(batch) < max_message_count
-            ):
+        try:
+            self._receive_context.set()
```

Comment:
Do we care that we're setting/clearing the threading Event regardless of whether we're actually using it (i.e. the callback is only replaced if prefetch is disabled)?
Probably not... doubtful that it would have much of a perf impact?

Reply:
tested multiple scenarios (both sync and async, iterator with prefetch1 vs 300, receive_messages with max_message_count 1 vs 300), compare the perf results with currently GAed lib and the perf impact is neglectable (almost no per diff from what I observed)